### PR TITLE
First time to inference POC

### PR DESCRIPTION
**Pull request by vurusovs**
POC for FTTI:
1. CLI contains `-m` and `-d` required arguments.
2. `runPipeline` encapsulates IE pipeline (or any other pipeline) and measures time of a pipeline. According to this design, pipeline.h may be redefined via CMake to generate executables with different pipelines.
3. Also `runPipeline` shouldn't throw exceptions and should handle it by itself
4. For time measurement implemented `Timer` class which measures time at start and finish of object's life cycle. For a correct use of `Timer` need to use `SCOPED_TIMER` macros

### CODE REVIEW

File: tests/time_tests/c++/pipeline.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+using namespace InferenceEngine;
+
+
+int runPipeline(float &duration) {
+    duration = []() {
+        auto t = std::chrono::high_resolution_clock::now();
+
+        Core ie;
+        CNNNetwork cnnNetwork = ie.ReadNetwork(FLAGS_m);
+        ExecutableNetwork exeNetwork = ie.LoadNetwork(cnnNetwork, FLAGS_d);
```
Comment (vurusovs) [inline tests/time_tests/c++/pipeline.h:14]: move globals to API of a pipeline to fix function arguments

File: tests/time_tests/c++/cli.h [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+/// @brief Define flag for showing help message <br>
+DEFINE_bool(h, false, help_message);
+
+/// @brief Declare flag for showing help message <br>
+DECLARE_bool(help);
+
+/// @brief Define parameter for set image file <br>
+/// i or mif is a required parameter
+DEFINE_string(i, "", input_message);
```
Comment (vurusovs) [inline tests/time_tests/c++/cli.h:78]: remove

File: tests/time_tests/c++/cli.h [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+/// @brief Define parameter for set model file <br>
+/// It is a required parameter
+DEFINE_string(m, "", model_message);
+
+/// @brief Define execution mode
+DEFINE_string(api, "async", api_message);
```
Comment (vurusovs) [inline tests/time_tests/c++/cli.h:85]: questionable

File: tests/time_tests/c++/cli.h [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+/// @brief device the target device to infer on <br>
+DEFINE_string(d, "CPU", target_device_message);
+
+/// @brief Absolute path to CPU library with user layers <br>
+/// It is a required parameter
+DEFINE_string(l, "", custom_cpu_library_message);
```
Comment (vurusovs) [inline tests/time_tests/c++/cli.h:92]: remove

File: tests/time_tests/c++/cli.h [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+/// @brief Define parameter for clDNN custom kernels path <br>
+/// Default is ./lib
+DEFINE_string(c, "", custom_cldnn_message);
```
Comment (vurusovs) [inline tests/time_tests/c++/cli.h:96]: remove

File: tests/time_tests/c++/cli.h [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+/// @brief Iterations count (default 0)
+/// Sync mode: iterations count
+/// Async mode: StartAsync counts
+DEFINE_uint32(niter, 0, iterations_count_message);
+
+/// @brief Number of infer requests in parallel
+DEFINE_uint32(nireq, 0, infer_requests_count_message);
```
Comment (vurusovs) [inline tests/time_tests/c++/cli.h:104]: questionable

File: tests/time_tests/c++/cli.h [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+DEFINE_string(l, "", custom_cpu_library_message);
+
+/// @brief Define parameter for clDNN custom kernels path <br>
+/// Default is ./lib
+DEFINE_string(c, "", custom_cldnn_message);
+
+/// @brief Iterations count (default 0)
+/// Sync mode: iterations count
+/// Async mode: StartAsync counts
+DEFINE_uint32(niter, 0, iterations_count_message);
```
Comment (vurusovs) [inline tests/time_tests/c++/cli.h:101]: questionable

File: tests/time_tests/c++/cli.h [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+#pragma once
+
+#include <string>
+#include <vector>
+#include <gflags/gflags.h>
+#include <iostream>
+
+/// @brief message for help argument
+static const char help_message[] = "Print a usage message";
```
Comment (vurusovs) [inline tests/time_tests/c++/cli.h:13]: make POSIX compatible CLI

Comment (vurusovs) [reply tests/time_tests/c++/cli.h:13]: To read: https://www.gnu.org/prep/standards/html_node/Command_002dLine-Interfaces.html

File: tests/time_tests/c++/pipeline.h [hunk 9]
```diff
@@ -0,0 +1,19 @@
+#include <chrono>
+#include <functional>
+
+#include <inference_engine.hpp>
+using namespace InferenceEngine;
+
+
+int runPipeline(float &duration) {
+    duration = []() {
+        auto t = std::chrono::high_resolution_clock::now();
```
Comment (vurusovs) [inline tests/time_tests/c++/pipeline.h:10]: create a class as MemCheckPipeline to encapsulate measures

File: tests/time_tests/c++/pipeline.h [hunk 10]
```diff
@@ -0,0 +1,19 @@
+#include <chrono>
+#include <functional>
+
+#include <inference_engine.hpp>
+using namespace InferenceEngine;
+
+
+int runPipeline(float &duration) {
```
Comment (vurusovs) [inline tests/time_tests/c++/pipeline.h:8]: func API should contain `model` and `device`

File: tests/time_tests/c++/pipeline.h [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+    duration = []() {
```
Comment (vurusovs) [inline tests/time_tests/c++/pipeline.h:9]: think about more than 1 measure in a pipeline

File: tests/time_tests/c++/pipeline.h [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+        auto t = std::chrono::high_resolution_clock::now();
+
+        Core ie;
```
Comment (vurusovs) [inline tests/time_tests/c++/pipeline.h:12]: think about handling exceptions in `runPipeline` (it's required if IPC will be implemented)

File: tests/time_tests/c++/main.cpp [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+
+int main(int argc, char **argv) {
+    if (!parseAndCheckCommandLine(argc, argv)) {
+        return -1;
+    }
+    statistics_path = "some_path"   // TODO: path should be unique for every run of executable
+    for (int i = 0; i < 3; i++) {
+        // TODO: aggregate 3 runs in one statistic
+        run_in_processes(1, runPipeline, statistics_path, FLAGS_m, FLAGS_d);
```
Comment (vurusovs) [inline tests/time_tests/c++/main.cpp:74]: move subprocess to Python part

File: tests/time_tests/c++/pipeline.h [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    // TODO: runPipeline may throw exceptions. In case of IPC need to handle exceptions via process status
+    TimePipeline(statistics_path);
+
+    Core ie;
+    CNNNetwork cnnNetwork = ie.ReadNetwork(model);
+    TimePipeline.record_measures("read_network");
+
+    ExecutableNetwork exeNetwork = ie.LoadNetwork(cnnNetwork, device);
+    TimePipeline.record_measures("first_time_inference");
+    return 0;
```
Comment (asomsiko) [inline tests/time_tests/c++/pipeline.h:72]: The usage shoudl be

int runPipeline(const std::string &model, const std::string &device) {
    Core ie;
    auto start_time = start_measure()
    CNNNetwork cnnNetwork = ie.ReadNetwork(model);
    record_time("read_network", start_time);
...
}

File: tests/time_tests/c++/pipeline.h [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * @brief Class response for encapsulating measure and measurements printing
+ *
+ * Current class measures only in scope of it's lifetime. In this case need
+ * to note that deletion of objects created before class creation may lead
+ * to negative values because of alignment on starting values.
+ * Also deletion of objects created in scope of class lifetime may decrease
+ * values computed on previous measure.
+ */
+class TimePipeline {
```
Comment (vurusovs) [inline tests/time_tests/c++/pipeline.h:17]: make it global so that some functions like `start/finish` to work with this class

Comment (vurusovs) [reply tests/time_tests/c++/pipeline.h:17]: think about implementing logic of few timers via one class

File: tests/time_tests/c++/timer.h [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    Timer(const Timer&) = delete;
+    Timer& operator=(const Timer&) = delete;
+public:
+    static Timer& Instance(){
+        static Timer _timer;
+        return _timer;
+    }
+
+    void startTimer(std::string timer_name) {
+        timers[timer_name].first = std::chrono::high_resolution_clock::now();
```
Comment (vurusovs) [inline tests/time_tests/c++/timer.h:24]: for a correct work need to add check that timer can't be rewritten

Comment (vurusovs) [reply tests/time_tests/c++/timer.h:24]: also need to add check that both measures (start and finish) have been set

File: tests/time_tests/c++/main.cpp [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+
+    return true;
+}
+
+
+int main(int argc, char **argv) {
+    if (!parseAndCheckCommandLine(argc, argv)) {
+        return -1;
+    }
+    runPipeline(FLAGS_m, FLAGS_d);
```
Comment (asomsiko) [inline tests/time_tests/c++/main.cpp:34]: Please check the return code. I also believe we should catch exceptions here. The goal is to ensure runPipeline does not abort main unexpectedly.

File: tests/time_tests/c++/pipeline.h [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+
+
+int runPipeline(const std::string &model, const std::string &device) {
+    start_timer("first_time_inference");
+
+    Core ie;
+
+    start_timer("read_network");
+    CNNNetwork cnnNetwork = ie.ReadNetwork(model);
+    finish_timer("read_network");
```
Comment (asomsiko) [inline tests/time_tests/c++/pipeline.h:16]: We should not remember the timer name. Scoped measure would help here.

File: tests/time_tests/c++/pipeline.h [hunk 19]
```diff
@@ -0,0 +1,22 @@
+#pragma once
+
+#include "timer.h"
+
+#include <inference_engine.hpp>
+using namespace InferenceEngine;
+
+
+int runPipeline(const std::string &model, const std::string &device) {
+    start_timer("first_time_inference");
```
Comment (asomsiko) [inline tests/time_tests/c++/pipeline.h:10]: Start time can return an object which stops the timer on destroy. This also enforces correct creation order if we want to measure "Core ie" construction. 
Please take a look at ScopedTask and OV_ITT_SCOPED_TASK design at openvino\itt\include\openvino\itt.hpp. It worth reusing those ideas.

File: tests/time_tests/c++/pipeline.h [hunk 20]
```diff
@@ -0,0 +1,22 @@
+#pragma once
```
Comment (asomsiko) [inline tests/time_tests/c++/pipeline.h:1]: The directory structure should follow the one in other openvino modules: module/src for internals and module/include for public includes. There is no "c++" directory

**Review by asomsiko — CHANGES_REQUESTED** [review]

File: tests/time_tests/c++/cli.h (hunk 3 above)
Comment (vurusovs) [reply tests/time_tests/c++/cli.h]: Removed all extra keys

File: tests/time_tests/c++/cli.h (hunk 2 above)
Comment (vurusovs) [reply tests/time_tests/c++/cli.h]: Done

File: tests/time_tests/c++/cli.h (hunk 5 above)
Comment (vurusovs) [reply tests/time_tests/c++/cli.h]: Done

File: tests/time_tests/c++/cli.h (hunk 7 above)
Comment (vurusovs) [reply tests/time_tests/c++/cli.h]: Removed

File: tests/time_tests/c++/cli.h (hunk 8 above)
Comment (vurusovs) [reply tests/time_tests/c++/cli.h:13]: GFlags provides unified way how specify command-line keys (so `--api`, `-api`, `-a`, `--a` is possible ways). Also I didn't find anything about positional arguments

File: tests/time_tests/c++/pipeline.h (hunk 9 above)
Comment (vurusovs) [reply tests/time_tests/c++/pipeline.h]: Implemented `Timer` class

File: tests/time_tests/c++/pipeline.h (hunk 11 above)
Comment (vurusovs) [reply tests/time_tests/c++/pipeline.h]: Done

File: tests/time_tests/c++/pipeline.h (hunk 14 above)
Comment (vurusovs) [reply tests/time_tests/c++/pipeline.h:20]: Implemented according to provided example

File: tests/time_tests/c++/pipeline.h (hunk 15 above)
Comment (vurusovs) [reply tests/time_tests/c++/pipeline.h]: Done

File: tests/time_tests/c++/timer.h (hunk 16 above)
Comment (vurusovs) [reply tests/time_tests/c++/timer.h]: Design of `Timer` is changed, so not required anymore

File: tests/time_tests/c++/pipeline.h (hunk 19 above)
Comment (vurusovs) [reply tests/time_tests/c++/pipeline.h]: Done

File: tests/time_tests/c++/pipeline.h (hunk 20 above)
Comment (vurusovs) [reply tests/time_tests/c++/pipeline.h:1]: Will be fixed after completing code review

File: tests/time_tests/c++/pipeline.h (hunk 18 above)
Comment (vurusovs) [reply tests/time_tests/c++/pipeline.h]: In this case we still need to handle timer name, but scoped measure doesn't require name specification more than once

File: tests/time_tests/c++/main.cpp (hunk 17 above)
Comment (vurusovs) [reply tests/time_tests/c++/main.cpp]: Done

File: tests/time_tests/c++/cli.h (hunk 4 above)
Comment (vurusovs) [reply tests/time_tests/c++/cli.h]: Done

File: tests/time_tests/c++/cli.h (hunk 6 above)
Comment (vurusovs) [reply tests/time_tests/c++/cli.h]: Done

File: tests/time_tests/c++/main.cpp [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    if (FLAGS_m.empty()) {
+        throw std::logic_error("Model is required but not set. Please set -m option.");
+    }
+
+    return true;
+}
+
+int runPipeline(const std::string &model, const std::string &device) {
```
Comment (vurusovs) [inline tests/time_tests/c++/main.cpp:31]: move to pipeline.h

File: tests/time_tests/c++/ie_pipeline.h [hunk 22]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include "timer.h"
+
+#include <inference_engine.hpp>
+using namespace InferenceEngine;
+
+
+void IEPipeline(const std::string &model, const std::string &device) {
```
Comment (asomsiko) [inline tests/time_tests/c++/ie_pipeline.h:9]: runPipeline

File: tests/time_tests/c++/ie_pipeline.h [hunk 23]
```diff
@@ -0,0 +1,19 @@
+#pragma once
```
Comment (vurusovs) [inline tests/time_tests/c++/ie_pipeline.h:1]: in general - update namings with CamelCase formatting

File: tests/time_tests/c++/ie_pipeline.h [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+
+
+void IEPipeline(const std::string &model, const std::string &device) {
+    auto fti_timer = start_timer("first_time_to_inference");
+
+    Core ie;
+
+    auto read_net_timer = start_timer("read_network");
+    CNNNetwork cnnNetwork = ie.ReadNetwork(model);
+    finish_timer(read_net_timer);
```
Comment (vurusovs) [inline tests/time_tests/c++/ie_pipeline.h:16]: try to get rid of `finish_timer`

File: tests/time_tests/c++/cli.h [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+
+/// @brief Declare flag for showing help message <br>
+DECLARE_bool(help);
+
+/// @brief Define parameter for set model file <br>
+/// It is a required parameter
+DEFINE_string(m, "", model_message);
+
+/// @brief device the target device to infer on <br>
+DEFINE_string(d, "CPU", target_device_message);
```
Comment (vurusovs) [inline tests/time_tests/c++/cli.h:35]: remove default value

File: tests/time_tests/c++/cli.h [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+
+/// @brief message for help argument
+static const char help_message[] = "Print a usage message";
+
+/// @brief message for model argument
+static const char model_message[] = "Required. Path to an .xml/.onnx/.prototxt file with a trained model or to a .blob files with a trained compiled model.";
+
+/// @brief message for assigning cnn calculation to device
+static const char target_device_message[] = "Optional. Specify a target device to infer on (the list of available devices is shown below). " \
+"Default value is CPU. Use \"-d HETERO:<comma-separated_devices_list>\" format to specify HETERO plugin. " \
```
Comment (vurusovs) [inline tests/time_tests/c++/cli.h:20]: remove default

File: tests/time_tests/c++/cli.h (hunk 8 above)
Comment (vurusovs) [reply tests/time_tests/c++/cli.h:13]: Decided to leave it as is

File: tests/time_tests/c++/pipeline.h (hunk 12 above)
Comment (vurusovs) [reply tests/time_tests/c++/pipeline.h]: Done

File: tests/time_tests/c++/main.cpp (hunk 13 above)
Comment (vurusovs) [reply tests/time_tests/c++/main.cpp]: Done

File: tests/time_tests/c++/pipeline.h (hunk 19 above)
Comment (vurusovs) [reply tests/time_tests/c++/pipeline.h]: Done via SCOPED_TIMER and specification of scopes

File: tests/time_tests/c++/pipeline.h (hunk 20 above)
Comment (vurusovs) [reply tests/time_tests/c++/pipeline.h]: Updated, please, review again

File: tests/time_tests/c++/main.cpp (hunk 21 above)
Comment (vurusovs) [reply tests/time_tests/c++/main.cpp]: Done

File: tests/time_tests/c++/ie_pipeline.h (hunk 22 above)
Comment (vurusovs) [reply tests/time_tests/c++/ie_pipeline.h]: Done

File: tests/time_tests/c++/ie_pipeline.h (hunk 23 above)
Comment (vurusovs) [reply tests/time_tests/c++/ie_pipeline.h]: Done for classes and functions, ignored for variables (but it still compatible with IE formatting)

File: tests/time_tests/c++/cli.h (hunk 25 above)
Comment (vurusovs) [reply tests/time_tests/c++/cli.h]: Done

File: tests/time_tests/c++/cli.h (hunk 26 above)
Comment (vurusovs) [reply tests/time_tests/c++/cli.h]: Done

File: tests/time_tests/c++/timer.h [hunk 27]
```diff
@@ -0,0 +1,37 @@
...
+
+    ~Timer(){
+        auto finish_time = std::chrono::high_resolution_clock::now();
+        float duration = std::chrono::duration_cast<std::chrono::microseconds>(
+                finish_time - start_time).count();
+        std::cout << name << ":" << duration << "\n"; // TODO: replace with writer
+    }
+};
+
+std::unique_ptr<Timer> start_timer(const std::string &timer_name) {
```
[the remark this answers is not in the document]
Comment (vurusovs) [reply tests/time_tests/c++/timer.h]: `start_timer` and `finish_timer` are removed due non-use. Done for any other items

File: tests/time_tests/c++/main.cpp [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+    try {
+        IEPipeline(model, device);
+    } catch (const InferenceEngine::details::InferenceEngineException& iex) {
+        std::cout << "Inference Engine pipeline failed with Inference Engine exception:\n" << iex.what();
```
[the remark this answers is not in the document]
Comment (vurusovs) [reply tests/time_tests/c++/main.cpp]: Yes, done

File: tests/time_tests/c++/timer.h [hunk 29]
```diff
@@ -0,0 +1,37 @@
...
+    return std::unique_ptr<Timer> (new Timer(timer_name));
+}
+
+void finish_timer(std::unique_ptr<Timer> &timer) {
```
[the remark this answers is not in the document]
Comment (vurusovs) [reply tests/time_tests/c++/timer.h]: Removed

File: tests/time_tests/c++/cli.h [hunk 30]
```diff
@@ -0,0 +1,48 @@
...
+
+/// @brief device the target device to infer on <br>
+DEFINE_string(d, "CPU", target_device_message);
+
+/**
+* @brief This function show a help message
+*/
+static void showUsage() {
+    std::cout << std::endl;
+    std::cout << "benchmark_app [OPTION]" << std::endl;
```
[the remark this answers is not in the document]
Comment (vurusovs) [reply tests/time_tests/c++/cli.h]: Renamed to TimeTests

File: tests/time_tests/c++/pipeline.h (hunk 10 above)
Comment (vurusovs) [reply tests/time_tests/c++/pipeline.h]: done

File: tests/time_tests/c++/pipeline.h (hunk 18 above)
Comment (vurusovs) [reply tests/time_tests/c++/pipeline.h]: Fixed

File: tests/time_tests/c++/ie_pipeline.h (hunk 24 above)
Comment (vurusovs) [reply tests/time_tests/c++/ie_pipeline.h]: done

**Review by asomsiko — APPROVED** [review]

**Pull request merged**